### DISCUSSION TITLE
WIZNET W5100s-EVB_PICO WebREPL Support

### DISCUSSION THREAD

**Discussion by IoTrees** [micropython/micropython#12852] [RP2040 / Pico] [answered]
Hello,

I have been trying to setup WebREPL on the WIZNET W5100s-EVB-PICO .
I was able to run the 'import webrepl_config' command, which modified my boot.py and created the 'webrepl_cfg.py' file in my main directory.

However, when I try to actually start the webREPL using the 'webrepl.start()' function, I get the following error:
```
File "webrepl.py", line 1, in start
  File "webrepl.py", line 1, in setup_conn
AttributeError: 'module' object has no attribute 'WLAN'
```

It kinda seems to me like the module is looking to start using WLAN as the network connection, but I am using the WIZNET chip instead of WLAN for the network interface.

Does anyone know how I could read the webrepl.py file on the board?

**Comment by jimmo** [comment, marked as answer]
This looks like a bug in webrepl. The offending code is here: https://github.com/micropython/micropython-lib/blob/master/micropython/net/webrepl/webrepl.py#L105

If you change those lines to query `network.LAN` instead, then copy webrepl.py to your device it will override the built-in version. We should fix this properly in webrepl (i.e. don't fail if it can't find `network.WLAN`, and also try `network.LAN`)

**Reply by jimmo** [reply]
Raised https://github.com/micropython/micropython-lib/issues/762

**Reply by IoTrees** [reply]
Thanks for the suggestion, I will try this tomorrow!
I am pretty new to the embedded python, is there anywhere special I need to put the updated file?

**Reply by jimmo** [reply]
Same place you put your main.py 

(In more detail, anywhere in `sys.path`, so `/lib` would also be fine)

**Reply by IoTrees** [reply]
I have been playing with this but still dont have it working.
I am not sure if I am understanding things correctly but here is what I have tried:

First, I tried changing the file as you mentioned, I found 'network,WLAN' in the file you linked, changed it to 'network.LAN' and uploaded but get a similar error:
```
  File "webrepl.py", line 107, in setup_conn          
AttributeError: 'module' object has no attribute 'LAN'
```

When I check to see the members of the 'network' module I get the list:
```
AP_IF           STA_IF          WIZNET5K        country
hostname        route        
```

I am thinking on this board, there is only the WIZNET5K class objects that can be used for network connections, dont know if this is correct.

From here, I tried changing 'network.LAN(i)' to 'network.WIZNET5K(i)', but get:
`TypeError: function missing 2 required positional arguments
`

I'm guessing the WIZNET5K board to be a little too different, so I tried changing the setup_conn function as follows, (mostly copied by the way the NIC is setup in my main.py):
```
def setup_conn(port, accept_handler):
    global listen_s
    listen_s = socket.socket()
    listen_s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    ai = socket.getaddrinfo("0.0.0.0", port)
    addr = ai[0][4]

    listen_s.bind(addr)
    listen_s.listen(1)
    if accept_handler:
        listen_s.setsockopt(socket.SOL_SOCKET, 20, accept_handler)
    # for i in (network.AP_IF, network.STA_IF):
    #     nic = network.LAN(i)
    spi = SPI(0, 2_000_000, mosi=Pin(19), miso=Pin(16), sck=Pin(18))
    nic = network.WIZNET5K(spi, Pin(17), Pin(20))  # spi,cs,reset pin
    nic.active(True)
    nic.ifconfig(('192.168.4.1', '255.255.255.0', '192.168.1.1', '8.8.8.8'))
    if nic.active():
        print("WebREPL server started on http://%s:%d/" % (nic.ifconfig()[0], port))
    else:
        print("Nic Not active", nic.active())
    return listen_s
```

This crashes my main.py, I think because I am trying to initialize the NIC twice, but after it crashes, I can run `webrepl.start()` and get:

```
WebREPL server started on http://192.168.4.1:8266/
Started webrepl in normal mode 
```

I cant get the client to connect, and I am not sure if it's even really starting, do you have any ideas?

**Reply by IoTrees** [reply]
EDIT: This does work! My problem was that my second network device I am using on my computer was not configured for the 192.168.4.XXX domain.

I changed the 'setup_conn' function to be:
```
def setup_conn(port, accept_handler):
    global listen_s
    listen_s = socket.socket()
    listen_s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    ai = socket.getaddrinfo("0.0.0.0", port)
    addr = ai[0][4]

    listen_s.bind(addr)
    listen_s.listen(1)
    if accept_handler:
        listen_s.setsockopt(socket.SOL_SOCKET, 20, accept_handler)
    # for i in (network.AP_IF, network.STA_IF):
    #     iface = network.LAN(i)
    spi = SPI(0, 2_000_000, mosi=Pin(19), miso=Pin(16), sck=Pin(18))
    iface = network.WIZNET5K(spi, Pin(17), Pin(20))  # spi,cs,reset pin
    iface.active(True)
    iface.ifconfig(('192.168.1.4', '255.255.255.0', '192.168.1.1', '8.8.8.8'))
    if iface.active():
        print("WebREPL server started on http://%s:%d/" % (iface.ifconfig()[0], port))
    else:
        print("iface Not active", iface.active())
    return listen_s
```

and now I am able to connect with my locally hosted client, it does respond to the correct password.
Im going to keep testing the featues but so far so good, thanks for your help!
